### PR TITLE
[Feature][connector][kafka] Support read Maxwell format message from kafka #4415

### PR DESCRIPTION
**Pull request by zhilinli123**
## Purpose of this pull request
issues:https://github.com/apache/incubator-seatunnel/issues/4415

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2](https://github.com/apache/incubator-seatunnel/tree/dev/docs/en/connector-v2)
  2. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  3. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Review by TyrantLucifer — DISMISSED** [review]
Please add e2e test case, thanks!

**Comment by zhilinli123** [comment]
> Please add e2e test case, thanks!

Ok, I'll finish it soon

**Comment by zhilinli123** [comment]
@ic4y  help run ci thanks

**Comment by zhilinli123** [comment]
@ic4y  help run ci thanks

**Comment by hailin0** [comment]
@zhilinli123 
Please check ci errors

**Comment by hailin0** [comment]
Please update your repository and merge the dev branch

**Comment by zhilinli123** [comment]
> Please update your repository and merge the dev branch

Done

**Comment by zhilinli123** [comment]
I will continue to complete it after waiting for ogg format pr merge!

**Comment by zhilinli123** [comment]
PTAL: @hailin0 @Hisoka-X thanks

**Comment by zhilinli123** [comment]
PTAL: @hailin0 @EricJoy2048 @Hisoka-X

**Review by EricJoy2048 — DISMISSED** [review]

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/maxwell/MaxWellJsonDeserializationSchema.java [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        return this.physicalRowType;
+    }
+
+    @Override
+    public void deserialize(byte[] message, Collector<SeaTunnelRow> out) {
+        if (message == null) {
+            return;
+        }
+        ObjectNode jsonNode = (ObjectNode) convertBytes(message);
+        assert jsonNode != null;
```
Comment (CheneyYin) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/maxwell/MaxWellJsonDeserializationSchema.java]: `assert` is turned off by default. I prefer throwing exceptions for non-test code.

Comment (zhilinli123) [reply seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/maxwell/MaxWellJsonDeserializationSchema.java]: From the Flink source code see received null message without any exception thrown, I think this is a format procedure does not throw exceptions will be more friendly
image

Comment (CheneyYin) [reply seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/maxwell/MaxWellJsonDeserializationSchema.java]: ACK

**Comment by zhilinli123** [comment]
PTAL: @hailin0  @Hisoka-X  thks

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/maxwellForMatIt/kafkasource_maxwell_cdc_to_pgsql.conf [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    topic = "maxwell-test-cdc_mds"
+    result_table_name = "kafka_name"
+    consumer.group = "maxwell_json_to_pg_consumer"
+    start_mode = earliest
+    schema = {
+      fields {
+           id = "int"
+           name = "string"
+           description = "string"
+           weight = "string"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/resources/maxwellForMatIt/kafkasource_maxwell_cdc_to_pgsql.conf]: test all datatypes on maxwell format? 

e.g.  float/double/decimal/date/time/timestamp/binary

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/maxwell/MaxWellJsonDeserializationSchema.java [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            out.collect(rowBefore);
+            out.collect(rowAfter);
+        } else if (OP_DELETE.equals(type)) {
+            SeaTunnelRow rowDelete = convertJsonNode(dataNode);
+            rowDelete.setRowKind(RowKind.DELETE);
+            out.collect(rowDelete);
+        } else {
+            if (!ignoreParseErrors) {
+                throw new SeaTunnelJsonFormatException(
+                        CommonErrorCodeDeprecated.UNSUPPORTED_DATA_TYPE,
```
Comment (CheneyYin) [inline seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/maxwell/MaxWellJsonDeserializationSchema.java]: `CommonErrorCodeDeprecated` has been annotated with `@Deprecated`. Please replace it with new `CommonErrorCode`.

**Comment by EricJoy2048** [comment]
Hi, @zhilinli123 

You submitted a great PR, but it hasn't been updated in a while, can you update it?

**Comment by zhilinli123** [comment]
> Hi, @zhilinli123
> 
> You submitted a great PR, but it hasn't been updated in a while, can you update it?

I am sorry that I have been a little busy with work recently. I will try my best to revise it as soon as possible

**Review by EricJoy2048 — DISMISSED** [review]
LGTM

PTAL @hailin0  @Carl-Zhou-CN  @TyrantLucifer

**Review by hailin0 — DISMISSED** [review]

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaFormatIT.java [hunk 4]
```diff
@@ -418,6 +425,26 @@ public void testFormatCompatibleCheck(TestContainer container)
...
+
+        LOG.info("====================== Check MaxWell======================");
+        // check ogg MaxWell to postgresql
+        Container.ExecResult execMaxWellResultToKafka =
+                container.executeJob("/maxwellFormatIT/kafkasource_maxwell_to_kafka.conf");
+        Assertions.assertEquals(
+                0, execMaxWellResultToKafka.getExitCode(), execMaxWellResultToKafka.getStderr());
+
+        Container.ExecResult execMaxWellResultToPg =
+                container.executeJob("/maxwellFormatIT/kafkasource_maxwell_cdc_to_pgsql.conf");
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaFormatIT.java:440]: check pg data row & column ？

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaFormatIT.java [hunk 5]
```diff
@@ -418,6 +425,26 @@ public void testFormatCompatibleCheck(TestContainer container)
...
+        Assertions.assertEquals(
+                0, execMaxWellResultToPg.getExitCode(), execMaxWellResultToPg.getStderr());
+
+        // Check Ogg
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaFormatIT.java]:
```suggestion
        // Check Maxwell
```

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaFormatIT.java (hunk 4 above)
Comment (zhilinli123) [reply seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaFormatIT.java:440]: KafkaFormatIT.checkMaxWellFormat done check

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaFormatIT.java [hunk 6]
```diff
@@ -95,6 +95,12 @@
 public class KafkaFormatIT extends TestSuiteBase implements TestResource {
 
     private static final Logger LOG = LoggerFactory.getLogger(KafkaFormatIT.class);
+
+    // ---------------------------Ogg Format Parameter---------------------------------------
```
Comment (Hisoka-X) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-kafka-e2e/src/test/java/org/apache/seatunnel/e2e/connector/kafka/KafkaFormatIT.java]:
```suggestion
    // ---------------------------Maxwell Format Parameter---------------------------------------
```

File: seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/maxwell/MaxWellJsonDeserializationSchema.java (hunk 1 above)
Comment (Hisoka-X) [reply seatunnel-formats/seatunnel-format-json/src/main/java/org/apache/seatunnel/format/json/maxwell/MaxWellJsonDeserializationSchema.java]: So we should change it just like flink?

**Comment by zhilinli123** [comment]
assert jsonNode != null;
Do you mean delete it?
@Hisoka-X

**Comment by Hisoka-X** [comment]
> ```
>     assert jsonNode != null;
> ```
> 
> Do you mean delete it? @Hisoka-X

yes

**Review by Hisoka-X — APPROVED** [review]

**Review by TaoZex — APPROVED** [review]
+1

**Pull request merged**